FlowCrypt/Functionality/Services/Key Services/KeyService.swift

```diff
@@ -110,17 +112,25 @@ final class KeyService: KeyServiceType {
 
     private func findKeyByUserEmail(keysInfo: [KeyInfo], email: String) throws -> KeyDetails? {
         let f: (KeyInfo) throws -> KeyDetails? = { keyInfo in
-            let parsedKeys = try self.coreService.parseKeys(
-                armoredOrBinary: keyInfo.`private`.data()
-            )
+            let parsedKeys = try self.keyParser(keyInfo.`private`.data())
 
             var keyDetails: [KeyDetails] = []
             parsedKeys.keyDetails.forEach { keyDetails.append($0) }
 
-            return keyDetails.first {
-                let emails = $0.pgpUserEmails
-                return emails.first == email || emails.contains(where: { $0 == email })
+            var foundKey: KeyDetails?
+            for key in keyDetails {
+                let emails = key.pgpUserEmails
+
+                if emails.first == email {
+                   return key
+                }
+
+                if foundKey == nil, emails.contains(where: { $0 == email }) {
+                    foundKey = key
+                }
             }
```

Comment:
This is what I wanted, though I wish the code was more obvious, like: (pseudocode)

```
if primaryEmailMatch = keyDetails.find({ $0.pgpUserEmails.first == email }) {
  return primaryEmailMatch
}
if anyEmailMatch = keyDetails.find({ $0.pgpUserEmails.contains(email) }) {
  return anyEmailMatch
}
return nil
```

That way, even someone who has no clue about what the code was for will figure it out quickly. As written now, it's less looping but more brain to understand (at least for me) even when I know what it's supposed to do and why.

Reply:
Idea here was to avoid double loop over the same data and also have only one `pgpUserEmails` since this is computational property and calculated (parsed) each time. No problem, will use your approach.

Reply:
Does Swift have any type for a cached (or calculated once) calculated property? In fact, the values in KeyDetails never change, can they be made immutable?